### PR TITLE
U/grelleum 4899 app provided table column extensions

### LINKED ISSUE #4899: App provided table column extensions
### As ...

P.D. - Plugin Developer

### I want ...

The ability to contribute extra columns to existing tables in core.

### So that ...

I can add additional data/context to a core table that is relevant to my app.

### I know this is done when...

I have a similar app API interface to filter extensions that allows me to register new columns to existing tables.

I expect to specify my columns as django_tables2/nautobot core Column classes.

Because columns are typically tightly coupled to data on the instance, it would be a good idea to provide a mechanism that allows me to optionally override/augment the queryset being used by the table. This way I could add a prefetch for my app's related model, etc.

### Optional - Feature groups this request pertains to.

- Plugins (and other Extensibility)

### Database Changes

_No response_

### External Dependencies

_No response_

### PR DESCRIPTION
**Pull request by grelleum**
# Closes #4899

# What's Changed
Added TableExtension class that allows app developers to add columns to core model tables.
- App can add one or more new columns to a table.
- App can add one or more, new or core columns, as default visible columns.
- App can remove one or more core default columns from the default visible columns.

# Screenshots
Unmodified Circuits table:

image

Circuits table modified by Example App.
- Added column `Provider ASN` as a default column.
- Removed column `Tenant` from the default columns.

image

# TODO
- [X] Explanation of Change(s)
- [X] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [X] Attached Screenshots, Payload Example
- [X] Unit, Integration Tests
- [X] Documentation Updates (when adding/changing features)
- [X] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: examples/example_app/example_app/table_extensions.py
Comment (grelleum) [reply examples/example_app/example_app/table_extensions.py:68]: I'll let the core team decide if this is useful or if it should be pulled out.

Comment (lampwins) [reply examples/example_app/example_app/table_extensions.py:68]: There are some examples that come to mind where this is useful, like when a custom column is derived from a specific custom relationship and thus the automatic logic is unable to generate a query optimization for that case. I have seen that a few times now.

Comment (glennmatthews) [reply examples/example_app/example_app/table_extensions.py:68]: Yeah - again, was trying to be clear that I totally understand the value of the feature in general. I was just curious about *this specific* queryset optimization for *this specific table*. :-)

File: nautobot/docs/development/apps/api/platform-features/table-extensions.md
Comment (grelleum) [reply nautobot/docs/development/apps/api/platform-features/table-extensions.md:17]: Interesting questions, I'll start with the last.

Because of the way this is implemented in the `CustomApp.NautobotAppConfig` class, it would be difficult to implement a weight system, because it would require knowledge of all modifications.  So it would need to be redesigned to accommodate that.   I can see the desire to have deterministic results.

There should be no issue if multiple apps remove columns from the same table. The `default_columns` is rebuilt using a comprehension that checks filters out any column names that are in the `remove_from_default_columns` list.  For example: When App 'A' is initialized it will remove column 'a' from the XYZTable, then when App 'B' is initialized column 'a' will no longer be in `default_columns`.  If App 'B' want to remove the same or different columns, it should not encounter any issues.

Last case: If one app wants to remove a column and another app want to add that same column.  The outcome would be determined based on the order the apps are initialized, with the last initialized app winning the battle.  I believe this scenario would not be common because:
1. If an app is adding a column to the table's `default_columns`, it is either a new column added by that app, and would be prefixed with the app_name.
2. If an app is adding a core column to the table's `default_columns`, it is likely not in the defaults and so it would not need to be removed.
3. If an app is removing a core column from the table's `default_columns`, it is likely already in the defaults and would not need to be added.

File: nautobot/extras/tests/test_plugins.py
Comment (grelleum) [reply nautobot/extras/tests/test_plugins.py:657]: Yes, that was the idea there - that it returns what is passed to it.  Really it is there for coverage.

File: nautobot/docs/development/apps/api/platform-features/table-extensions.md
Comment (lampwins) [reply nautobot/docs/development/apps/api/platform-features/table-extensions.md:11]: For now, let's just focus on the models, and thus the list views. I'd be concerned with tightly coupling this to specific tables, and issues with this breaking as updates occur. Perhaps it can be implemented as an advanced option in the future, but not needed now.

File: nautobot/docs/development/apps/api/platform-features/table-extensions.md
Comment (lampwins) [reply nautobot/docs/development/apps/api/platform-features/table-extensions.md:3]: `develop`/2.3

File: examples/example_app/example_app/table_extensions.py
Comment (grelleum) [reply examples/example_app/example_app/table_extensions.py:68]: It is in the example_app and it's not really an optimization, just a demo how to use it.  I get that the examples should be meaningful, but also they should be simple, so I think I errored on the side of simplicity.

File: nautobot/extras/tests/test_plugins.py
Comment (grelleum) [reply nautobot/extras/tests/test_plugins.py:703]: I can modify this to change the behavior.  It is based on the `FilterExtension` implementation, which will fail in a similar way.

File: nautobot/docs/development/apps/api/platform-features/table-extensions.md
Comment (grelleum) [reply nautobot/docs/development/apps/api/platform-features/table-extensions.md]: I've updated this to 2.3.9.

File: nautobot/extras/tests/test_plugins.py
Comment (grelleum) [reply nautobot/extras/tests/test_plugins.py]: today I learned ... `self.assertLogs`

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:278]: Add `if table_extension.table_columns` check here?

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:582]: Not opposed to extra error handling, just curious - with the requirement that the column name include the app name as a prefix, this should pretty much "never" occur, no?

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:576]: Does this need to be a local import here, or could it be moved to the top of the file?

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:592]: Hm. Does this work for both old-style `ObjectListView` as well as `NautobotUIViewSet`? I'm assuming so since the example_app is extending one of each - but I'm a bit concerned in the NautobotUIViewSet case since the class-level `queryset` attribute is shared across all sub-views under the ViewSet. I wonder if in that case it **might** be safer to actually wrap/monkey-patch the `view.alter_queryset()` functions to just call `table_extension.alter_queryset()` when specifically handling a list request?

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [inline nautobot/extras/plugins/__init__.py:608]: It *might* be nice to log a debug message if `remove_from_default_columns` includes any items that aren't actually in the table's default_columns? Definitely no higher than `debug` though.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:592]: I was concerned about the `NautobotUIViewSet`, which is why I added the code to extend the Circuit table.  This is a good point and I'll look into modifying this behavior.

Comment (glennmatthews) [reply nautobot/extras/plugins/__init__.py:592]: Much appreciate your thoroughness! :-)

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:576]: This needs to be a local import.  Otherwise we get this when trying to start the server or workers.
```
celery_worker-1  | Traceback (most recent call last):
celery_worker-1  |   File "/usr/local/bin/nautobot-server", line 3, in <module>
celery_worker-1  |     from nautobot.core.cli import main
celery_worker-1  |   File "/source/nautobot/core/cli/__init__.py", line 17, in <module>
celery_worker-1  |     from nautobot.extras.plugins.utils import load_plugins
celery_worker-1  |   File "/source/nautobot/extras/plugins/__init__.py", line 7, in <module>
celery_worker-1  |     import django_tables2
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django_tables2/__init__.py", line 19, in <module>
celery_worker-1  |     from .tables import Table, table_factory
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django_tables2/tables.py", line 186, in <module>
celery_worker-1  |     class Table(metaclass=DeclarativeColumnsMetaclass):
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django_tables2/tables.py", line 26, in __new__
celery_worker-1  |     attrs["_meta"] = opts = TableOptions(attrs.get("Meta", None), name)
celery_worker-1  |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django_tables2/tables.py", line 121, in __init__
celery_worker-1  |     DJANGO_TABLES2_TEMPLATE = getattr(
celery_worker-1  |                               ^^^^^^^^
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 102, in __getattr__
celery_worker-1  |     self._setup(name)
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 89, in _setup
celery_worker-1  |     self._wrapped = Settings(settings_module)
celery_worker-1  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^
celery_worker-1  |   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 217, in __init__
celery_worker-1  |     mod = importlib.import_module(self.SETTINGS_MODULE)
celery_worker-1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_worker-1  |   File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
celery_worker-1  |     return _bootstrap._gcd_import(name[level:], package, level)
celery_worker-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_worker-1  | ModuleNotFoundError: No module named 'nautobot_config'
```

**Comment by itdependsnetworks** [comment]
Can you update docs here as well? https://github.com/nautobot/nautobot/blob/044490de8c4e88a09be11cdd0c31ccc7c19140bf/nautobot/docs/development/apps/api/setup.md?plain=1#L11

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [reply nautobot/extras/plugins/__init__.py:576]: Ack. Didn't know that importing tables2 forced evaluation of the Django settings.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:582]: Pretty much it should not occur - the app name prefix should provide a namespace to avoid that. Really the only conceivable way that could happen would be if the app name was the same as a base model.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py]: I've refactored this code and added debug logging for when a column name is missing.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py]: I've refactored this code and added debug logging for when a column name is missing.

**Comment by grelleum** [comment]
> Can you update docs here as well?
> 
> https://github.com/nautobot/nautobot/blob/044490de8c4e88a09be11cdd0c31ccc7c19140bf/nautobot/docs/development/apps/api/setup.md?plain=1#L11

Added!

File: nautobot/extras/tests/test_plugins.py
Comment (grelleum) [reply nautobot/extras/tests/test_plugins.py]: I have changed this to log an error rather than raise an exception as recommended.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:622]: I am still pending on this.

**Comment by glennmatthews** [comment]
> @glennmatthews Do you think there would be a benefit if I moved all the TableExtension code to a `table_extensions.py` module, exposed the two functions needed by `__init__` and imported that?

I don't have a strong opinion but moving stuff out of `__init__.py` in the long run would be nice. Maybe move to the existing `tables.py` rather than creating a new module?

File: nautobot/extras/plugins/__init__.py
Comment (glennmatthews) [reply nautobot/extras/plugins/__init__.py]: New approach looks good.

**Comment by grelleum** [comment]
>  Maybe move to the existing `tables.py` rather than creating a new module?
Importing `tables` into `__init__` prevents Nautobot from starting.  `ModuleNotFoundError: No module named 'nautobot_config'`.
I'll skip it and focus on isolating the queryset for the list view.

**Comment by glennmatthews** [comment]
> > Maybe move to the existing `tables.py` rather than creating a new module?

> Importing `tables` into `__init__` prevents Nautobot from starting.  `ModuleNotFoundError: No module named 'nautobot_config'`.

Ack, same problem as before. thanks for trying it.

File: nautobot/extras/plugins/__init__.py
Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:622]: I am unsure of a good way to patch this in the UIViewSet that targets just the list view queryset.  I think I would need to patch either the `.as_view` or the `.get_queryset` method, however, I would sill need to identify how the method is being called to know if it is for a list view or not.  I'm not saying it can't be done, but I'm not sure what the effort would be and if it is worth mucking that deep into the internals.

Comment (glennmatthews) [reply nautobot/extras/plugins/__init__.py:622]: Fair enough. I think maybe for now, we can leave it as is with a TODO comment and see if it causes any issues?

Comment (grelleum) [reply nautobot/extras/plugins/__init__.py:624]: Good deal - I've added a TODO on this.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**
